### ISSUE TITLE
Initial popping sound in upward compressor

### ISSUE THREAD

**Issue by Digitalone1** [sadko4u/lsp-plugins#121]
Hello, I have a [preset](https://github.com/Digitalone1/PulseEffects-Presets) in [PulseEffects](https://github.com/wwmm/pulseeffects) relying on lsp compressor set in upward mode.

As a user stated in an [issue](https://github.com/Digitalone1/PulseEffects-Presets/issues/1), the plugin produces a popping sound at the start and the only way to reduced it is to decrease the attack parameter at the minimum, which makes it softer but does not eliminate it completely. This can be noticed switching from mute to unmute while watching something.

I thought this was caused by the attack parameter that is increasing the gain at the start producing an output above 0 db for a very short amount of time. It it's so, setting attack to 0 ms should eliminate it, but that's not occurring.

It can be eliminated only setting the feed-forward type, but unfortunately the upward mode is less aggressive in this configuration and that's not what I'm looking for.

I also tried to play with knee setting different values, but it seems not affecting the upward mode and does not change this behavior.  

Why is that occurring in feed-back type and is there another way to eliminate it completely? Thanks.

**Comment by sadko4u** [comment]
The answer comes from the understanding of how the upward compressor does work: the upward compressor amlifies silent signals and does nothing with loud signals.

The characteristics of the output amplification gain of the compressor depend on the compressor's signal envelope value which is reactive and controlled by the Attack value. When the attack is low, envelope changes quickly. When attack is huge, envelope changes slowly. It is shown on the picture below (sorry for handwriting but it was easier to draw by hand):

![pop-effect](https://user-images.githubusercontent.com/12834464/85946219-87c8f000-b94b-11ea-8fb9-ac10cb91d8ae.png)

Here we see two cases:
- Case # 1 - when Attack value is very small.
- Case # 2 - when Attack value is big.

We try to simulate the situation when the playback triggers: at one instant the input signal significantly and immediately raises from -inf dB to around 0 dB.
In Case # 1 the envelope raises immediately and gain becomes reduced immediately, too, because of small attack time. So we get no audible pop because compressor does not amplify the signal.
In Case # 2 the envelope raises slowly, so slowly the gain becomes reduced, and we get an initial pop which is additionally amplified for some amount of time.

You're not the first person who asks me about pops for the upward compressor.
Here's already related issue: https://github.com/sadko4u/lsp-plugins/issues/72

To not to trigger loud pops, the audio playback should not issue a pop, it should perform a short fade-in when the playback starts. But I believe I need to implement some de-popper which will control the input signal for loud pops and smooth them for short amount of time (50 - 100 ms).

**Comment by Digitalone1** [comment]
Thanks for the explanation.

In the related issue you recommend to add a gate. It's interesting that I already included a gate before the upward compressor, but the pop remains. Maybe I should change its parameters to perform a short fade-in which is not occurring now. I'll test and let you know.

**Comment by Digitalone1** [comment]
Unfortunately the gate in PulseEffects does not help in this case. When the playback is paused for longer than a certain amount of time, the gain reduction is lost. Then, after unmuting, the pop reappears. 

At the moment there's nothing to perform a fade-in in PulseEffects after the unmute, so I hope some workaround can be found inside the upward compressor.

**Comment by sadko4u** [comment]
You also may use lookahead option as a workaround. In this case the compressor will already react before the playback starts and the pop will be compensated. Try to set 5-10 ms of lookahead and fast, non-zero attack.

**Comment by Digitalone1** [comment]
Setting a lookahead seems better.

**Comment by Digitalone1** [comment]
So I tested it in these days and can confirm that setting a very short non-zero attack with a short sidechain lookahead is resolving the issue. I can hear a little pop, but it's very soft. Sometimes I can hear it also when the PE global bypass is set, so the only way to not trigger it is to set a short fade-in in both cases. Anyway it'is not a compressor issue.

But, since the pop is amplified when the attach is not short, everyone is forced to set a short attack and that's a great limitation. Maybe something can be done to workaround this situation? In example, the knee setting can be used to reduce this behavior.

In downward mode, knee 0 is applying the reduction instantly, while higher the value smoother the reduction. But I see that knee setting is not applied in upward mode. Maybe I'm wrong, but I hear no difference setting different values. I was hoping that setting the highest knee would smooth the amplification reducing the pop even at larger attack.

**Comment by sadko4u** [comment]
The short explanation of the Lookahead feature. It just works as a delay to the signal channel. So we compute the envelope of the signal **before** it actually will be but delayed audio signal makes the compressor to start working earlier the pop occurs. So that's why you get a smoothed pop.
Anyway since this topic is still open, I'll work on the solution.

**Comment by sadko4u** [comment]
@Digitalone1 
I've developed a new plugin called **Surge Filter**. It automatically detects the loudness surge and performs fade-in for configurable period of time. Are you available to test this plugin before the upward compressor?
If it helps you, I'll build-in part of this module into compressor plugins.

@v-fox
I think this also will be interesting for you since you've had problems with upward compressor.

Changes are available in **github-issue-121** branch.

**Comment by v-fox** [comment]
@sadko4u Build failed for me with:
```
[   85s] PHP Fatal error:  require(): Failed opening required './manuals/plugins/surge_filter.php' (include_path='.:/usr/share/php7:/usr/share/php7/PEAR') in /home/abuild/rpmbuild/BUILD/lsp-plugins-1.1.22+69~20200708T011000.166/.build/src/doc/index.php on line 63
[   85s] make[3]: *** [Makefile:151: /home/abuild/rpmbuild/BUILD/lsp-plugins-1.1.22+69~20200708T011000.166/.build/html/html/plugins/surge_filter_mono.html] Error 255
```

**Comment by sadko4u** [comment]
@v-fox 
Added missing documentation stub file (documentation is still not written).

**Comment by v-fox** [comment]
@sadko4u Tried it but I can't figure out a config comparable to my gate config from #72 It either ineffectual or too aggressive. Control knobs are all over the place and I'm not sure that they even show real numbers. For now I've stopped at this but it still pops:
![lsp-plugins-surge-filter-stereo](https://user-images.githubusercontent.com/362763/86862802-5d152f00-c0e3-11ea-9db7-0cd4c79c7221.png)
Testing on talking youtubers by seeking with arrow-keys or starting and stopping. Note that volume in such videos often non-normalized and non-compressed, so it usually varies well below 0 (which is why I've set such big preamp in gate) with occasional spike to the point of clipping.

**Comment by Digitalone1** [comment]
> Are you available to test this plugin before the upward compressor?
I only use PulseEffects and since there's no specific plugin in the pipeline, I think there's no way. Never used Ardour or other similar application.

> If it helps you, I'll build-in part of this module into compressor plugins.
I can build a custom lsp-plugin package on my system and test inside PulseEffects.

**Comment by sadko4u** [comment]
@v-fox 
I see on the time graph that it detects spikes:

Please also try longer **Fade in** settings.

![image](https://user-images.githubusercontent.com/12834464/86896469-ae86e380-c10e-11ea-9345-ba3c3f149ae8.png)

**Comment by v-fox** [comment]
@sadko4u Yes, I also have eyes that are capable of differentiating perpendicular lines. It doesn't address my points though. I've already said that I was unable to find adequate parameters, it either pops (same narrow spike) or cuts too much real sound (weirdly shaped curve that has little to do with release and fade-in numbers). I've wiggled controls all over the place. So if even you don't know what would be adequate then it's all guesswork for me. Look at my gate parameters from #72, how abnormal they are and even that is far from perfect (pops are also not entirely gone but settings on which they were also similarly silenced real sounds). This is even more unintuitive.

Such limited threshold range doesn't make it easier. I'm always asking you to allow ranges of at least full 24 bit encoding (144 dB or something like that) everywhere but you newer want to. On tracks like that, where peaks are in -24/-36 range, "dips" are even lower so upward compressor works extra-hard on amplifying quiet background sounds. When those sounds happen to be at threshold line, this plugin starts violently clicking. But gate has similar problem, it only masks it with release time which actually works as expected. Neither is good because threshold should be far from level of lowest desirable discernable sound (+ account for summary headroom of all potential stages of pipeline with 3-12 dB each), yet they both have limited threshold range.

**Comment by sadko4u** [comment]
I think I need to rework modules a bit. I have an idea how 'surge protection' should work in a bit another way, respectively to the real RMS level of the audio signal, not 0 dBFS RMS.

**Comment by sadko4u** [comment]
@v-fox @Digitalone1 

I've pushed many new changes to the branch (and rebased it to actual **devel** branch).
Please try new surge filter. I've tested it with aggressively set up upward compressor and it seems that pops should be gone.

**Comment by Digitalone1** [comment]
Can you include it in the compressor itself?

**Comment by sadko4u** [comment]
@Digitalone1 
I'm currently thinking on it. If I add the surge protector, I need to add some controls to the UI and think about how not to bloat the UI...

**Comment by sadko4u** [comment]
Anyway I need feedback from @v-fox about does this plugin solve his needs or not.

**Comment by v-fox** [comment]
@sadko4u So, I've built the new version:
![lsp-plugins-surge-filter-stereo_v2](https://user-images.githubusercontent.com/362763/87084101-1dffee80-c247-11ea-9d6d-3441c0cfaef8.png)

Seems to work quite nicely indeed, better even than the gate ! But:
* I haven't noticed any difference with various modes.
* Threshold range is adequate now but, despite usage of 32-bit float by JACK, still is narrower than default 24-bit encoding of pretty much all modern DACs/ADCs and fancy FLACs & movie DTS tracks, which cheap decent closed-back headphones should be able to reproduce, which leaves no room for headroom to guarantee having no theoretical loss of range even at max.
* Graph is even more limited with mere 60 dB.
* It's the only LSP plugin that has graph in window thumbnail when minimized in KDE which leads me to believe that it does not stop rendering when minimized.
* Having that functionality in compressor would allow to decrease number of opened plugins which is nice but I also have no idea where it could be put.
  * But compressor, and pretty much all LSP in general, has same limitations in knob & graph range.
  * And I'm also would be glad to see dual S-mode for the compressor with simultaneous upward compression for bottom and downward compression on top which would turn it into the most monstrous audio plugin ever. But that functionality could be replicated with second compressor or limiter after the first one. It would only make sense if integration would decrease CPU use and/or latency.

**Comment by sadko4u** [comment]
> I haven't noticed any difference with various modes.

Anyway they are. Try to set long fade in and you'll hear the difference. Gaussian is more aggressive while cubic and cosine are good compromises.

> Threshold range is adequate now but, despite usage of 32-bit float by JACK, still is narrower than default 24-bit encoding

I think you're not correct here. 32-bit floating point consists of 1 sign bit, 8 bit of exponent and 23 bits of fraction part.
All of this allows to represent a 25-bit number without any losses (1 bit of fraction is always considered to be 1 and that's why omitted in the encoding).
https://en.wikipedia.org/wiki/Single-precision_floating-point_format

I don't see any benefit of using threshold of -144 dB. Trying to amplify further such low signals will give high SNR value and probably impossible to listen the final track without frustrations.

> Graph is even more limited with mere 60 dB.

> It's the only LSP plugin that has graph in window thumbnail when minimized in KDE which leads me to believe that it does not stop rendering when minimized.

It's not the only plugin. Look at trigger plugin or spectrum analyzer plugin. They all render their main graphs into an application icon.

> And I'm also would be glad to see dual S-mode for the compressor with simultaneous upward compression for bottom and downward compression on top which would turn it into the most monstrous audio plugin ever.

You may build your own compressor by using Dynamic Processor Plugin:

![image](https://user-images.githubusercontent.com/12834464/87087873-872e3480-c23c-11ea-9761-95ba961ccc7f.png)

Here's short demo:

https://www.youtube.com/watch?v=rvEhP_PRtzk

**Comment by v-fox** [comment]
> > I haven't noticed any difference with various modes.
> 
> Anyway they are. Try to set long fade in and you'll hear the difference. Gaussian is more aggressive while cubic and cosine are good compromises.

Ah, yeah. I though that I've tested it with ~500ms but, apparently, not and now I see the difference. Still going to keep it short.

However, now I've also noticed that sudden stops produce clicking sound. Probably because compressor also amplifies sudden pauses and plugin does not soften that. Lookahead might be required ? In that case it would be better to combine lookahead of compressor with that instead of increasing delay by having both.

> > Threshold range is adequate now but, despite usage of 32-bit float by JACK, still is narrower than default 24-bit encoding
> 
> I think you're not correct here. 32-bit floating point consists of 1 sign bit, 8 bit of exponent and 23 bits of fraction part.
> All of this allows to represent a 25-bit number without any losses (1 bit of fraction is always considered to be 1 and that's why omitted in the encoding).
> https://en.wikipedia.org/wiki/Single-precision_floating-point_format
> 
> I don't see any benefit of using threshold of -144 dB. Trying to amplify further such low signals will give high SNR value and probably impossible to listen the final track without frustrations.

https://www.sounddevices.com/32-bit-float-files-explained/ - this says that 32-bit float encoding somehow has 1528 dB of range. I don't know either way but that besides the point of cutting out parts of original source for no reason by artificially limiting knobs and buttons.

Professionally mastered music and movies would not have too much of annoying noise as background and you're still ignoring that sum of headroom of all preceding plugins/processors may shift volume lower, sometimes even tens of dB which will cut off not just few tens dB of difference with lowest knob value from original but also some more above that.

And even without any amplification it would be nice to see actual noise levels of recordings to set threshold without guessing where they are.

> > Graph is even more limited with mere 60 dB.
> 
> > It's the only LSP plugin that has graph in window thumbnail when minimized in KDE which leads me to believe that it does not stop rendering when minimized.
> 
> It's not the only plugin. Look at trigger plugin or spectrum analyzer plugin. They all render their main graphs into an application icon.

Not when they are minimized… actually, they don't render graph-icon for me in KDE even when they are visible (it's just static blue "LSP" logo) and taskbar thumbnail is rendered. Maybe it's a bug in kwin/plasma or the rest of them.

> > And I'm also would be glad to see dual S-mode for the compressor with simultaneous upward compression for bottom and downward compression on top which would turn it into the most monstrous audio plugin ever.
> 
> You may build your own compressor by using Dynamic Processor Plugin

But how is that any better than launching 2 of them with different configurations ?

**Comment by sadko4u** [comment]
> Ah, yeah. I though that I've tested it with ~500ms but, apparently, not and now I see the difference. Still going to keep it short.

I think I need to reduce the upper bound up to 1-2 seconds, not more.

> However, now I've also noticed that sudden stops produce clicking sound. Probably because compressor also amplifies sudden pauses and plugin does not soften that. Lookahead might be required ? In that case it would be better to combine lookahead of compressor with that instead of increasing delay by having both.

Plugin does nothing with stops. It just triggers fade-in processing when the signal raises above the specified threshold. 
The upward compressor should not amplify at the moment when the stop occurs.

Anyway, proper fade-outs require some amount of lookahead and will introduce some latency.

**Comment by sadko4u** [comment]
Added fade-out support by the surge filter, pushed changes to the **github-issue-121** branch: b8500dfb473c2b1c2adf7e81c118f81b47335add

![image](https://user-images.githubusercontent.com/12834464/87210860-c4b7be00-c31f-11ea-94b8-b8f65a0c90f3.png)

Please check

**Comment by v-fox** [comment]
@sadko4u 
* As I suspected, hidden lookahead delay is equal to fade-out time which drops 100ms unto entire pipeline which is completely incompatible with real-time use. I doubt that there are many people who would actually measure pipeline delay like I do, so only few would suspect what's happening and where. I would suggest to put big red warning labels on all delay-inducing parameters everywhere and put their defaults on bare effective minimum. I think that 16ms of total system delay is the point that, ideally, should not be crossed, >32ms is the worst case scenario, 64-128ms can be allowed only for heavy real-time mixing with thousands of sources and whole bunch of processing. Anything beyond is Cthulhu-level madness.
* It can't be set below 1 ms or disabled.
* It seems to be adds to release time, or something else is at play, because plugin skips points where it obviously should have triggered. Such as quick pause/unpause or seeking a video with arrow keys. But sometimes it just goes straight at silent parts of tens of ms, similar to those that would trigger it other times.
* Again, because view is limited to arbitrary 60 dB noise floor, that may or may not be preventing it from triggering at some situations, is invisible. As is the full curve of the plugin. It can't even be forced to go up with gain knobs because those are also limited.
* Not sure what "attack" supposed to do.

**Comment by sadko4u** [comment]
> As I suspected, hidden lookahead delay is equal to fade-out time which drops 100ms unto entire pipeline which is completely incompatible with real-time use.

Of course, there are no psychics that can predict the further behaviour of the signal except for just implementing delay to properly draw the fade-out in the past.

> I would suggest to put big red warning labels on all delay-inducing parameters everywhere and put their defaults on bare effective minimum

Changed the color of the knob to red

> It can't be set below 1 ms or disabled

Setting of 0 ms is now default

> Again, because view is limited to arbitrary 60 dB noise floor, that may or may not be preventing it from triggering at some situations, is invisible.

Extended graph up to -144 dB.

> Not sure what "attack" supposed to do.

The envelope controls the plugin to trigger fade in or fade out. Now the envelope is additionally displayed at the graph. The 'Attack' and 'Release' controls show how smooth the envelope is.

**Comment by sadko4u** [comment]
Pushed latest changes: 7abb0ebb09b8760c38ff9781522e8c4966d6322f

**Comment by Digitalone1** [comment]
I'm getting used to the short attack and little lookahed, so don't think I will need this surge filter. Anyway, if you are adding it to the compressor, I'll make the necessary changes to PulseEffects code to expose it inside the application GUI. Thanks for your work.

**Comment by v-fox** [comment]
> > As I suspected, hidden lookahead delay is equal to fade-out time which drops 100ms unto entire pipeline which is completely incompatible with real-time use.
> 
> Of course, there are no psychics that can predict the further behaviour of the signal except for just implementing delay to properly draw the fade-out in the past.

> > I would suggest to put big red warning labels on all delay-inducing parameters everywhere and put their defaults on bare effective minimum
> 
> Changed the color of the knob to red

Sure but that's I nasty surprise for someone who wouldn't immediately figure out that `lookahead`=`latency` and `fade-out time`=`lookahead`. Maybe it should also be labelled as "lookahead fade-out time" or something like that.

> > It can't be set below 1 ms or disabled
> 
> Setting of 0 ms is now default

> I'm getting used to the short attack and little lookahead, so don't think I will need this surge filter. Anyway, if you are adding it to the compressor, I'll make the necessary changes to PulseEffects code to expose it inside the application GUI. Thanks for your work.

Nice. Although, integrating in the compressor would allow sharing lookahead buffer and exclude additional latency, right ?

> > Again, because view is limited to arbitrary 60 dB noise floor, that may or may not be preventing it from triggering at some situations, is invisible.
> 
> Extended graph up to -144 dB.

Even nicer ! Would be cool to get that in `lsp-plugins-spectrum-analyzer-x4` to also see real noise-floor. It also has some weird idea about volume levels that other plugins do not seem to share. Strange how it doesn't have 1x1+1x1 comparison mode too. Calf has some interesting modes in its visualizer, especially "stereo wideness square" and "zoom" slider… which, to be fair, is also limited in its range. Its lack of colour palettes and differentiation sucks, though. Wouldn't auto-rainbow for frequency range and desaturation/transparency for volume range inside hard outline for a single graph and coloured intersection filling for differentiation graph be dope ?

New, big graph revealed weird things going below 60 dB: on what supposed to be sharp pause some kind of slow fade-out is happening even with all preceding plugins on bypass. I though that maybe Firefox and/or PulseAudio were doing something but with Strawberry, with JACK backend and internal fade-out disabled, result is the same. That prevents it from triggering:
![lsp-plugins-surge-filter-stereo_weird-preceding-fade-out](https://user-images.githubusercontent.com/362763/87238566-7d175c00-c41d-11ea-8c69-6278d7eace0a.png)

While checking that out I've found a bug in `lsp-plugins-latency-meter`: it enables feedback on bypass which immediately saturates whole stream with unholy noise.

> > Not sure what "attack" supposed to do.
> 
> The envelope controls the plugin to trigger fade in or fade out. Now the envelope is additionally displayed at the graph. The 'Attack' and 'Release' controls show how smooth the envelope is.

Does fade-out trigger only on original signal curve or on post fade-in signal ?

When fade-in and fade-out threshold are different it randomly mutes stream on some pause/unpause cycles until next pause/unpause:
![lsp-plugins-surge-filter-stereo_stuck-configuration](https://user-images.githubusercontent.com/362763/87238562-74268a80-c41d-11ea-988b-3534f2d2e8e3.png)

**Comment by sadko4u** [comment]
Rebased the branch to actual **devel** branch. New limiter is now available in this branch.

**Comment by Digitalone1** [comment]
> Nice. Although, integrating in the compressor would allow sharing lookahead buffer and exclude additional latency, right ?

I don't know, these are technical stuff regarding Pulseaudio that @wwmm might be aware.

**Comment by wwmm** [comment]
As we have bigger latency sources I wouldn't worry to much about this.

**Comment by sadko4u** [comment]
@v-fox 
I've submitted new changes to the surge filter plugin. Now it tracks RMS in order to know the proper start point of the fade-out event. Additionally, time protection has been added. Consider the following case:
1. We set fade-out threshold greater than fade-in threshold.
2. We start playback. The signal envelope raises over the fade-in threshold. The actual fade-in only starts when the signal envelope is greater than fade-in threshold for the specified fade-in delay. This saves us from short spikes.
3. Same for fade-out: the fade-out triggers immediately if the envelope is below the fade-out threshold. But the system can not fade in again for less than specified fade-out delay.

Anyway, now the device works more reliable for me. Please check

**Comment by v-fox** [comment]
@sadko4u Yep, looks nice now.
I still don't understand where all those sub-72 dB slow curves on pause of every source come from but this version seems to allow work around them. Why RMS shown red since it doesn't seem to add latency in lsp-plugins-latency-meter ? What its tuning supposed to affect ?

> As we have bigger latency sources I wouldn't worry to much about this.

@wwmm Latency sources like what ? Even beefy limiter with big lookahead doesn't seem to compare to a decent fade-out. Realtime sources would usually target from 1 video-frame (≤16 ms) to 64 ms and non-realtime - from 100ms to 2s. So, I think that every ms counts in realtime use.

**Comment by sadko4u** [comment]
> Why RMS shown red since it doesn't seem to add latency in lsp-plugins-latency-meter ? What its tuning supposed to affect ?

RMS also affects the latency since RMS value requires to be computed for a period of time. Now consider the following situation: the audio suddenly disappears. The RMS value will fall to 0 only after the RMS time. So we need to draw the fade out **before** the RMS value has fallen below the fade-out threshold. So the easiest way to do this, just to start draw fade-out **RMS** milliseconds before the event appeared. But also our fade out has non-zero finite length. So we need additionally to draw fade-out within **Time** milliseconds. So actual latency we get is **RMS** + **Fade-out time** in milliseconds.

**Comment by wwmm** [comment]
I did not explain myself very well. As I told in my latest answer on the other thread a GStreamer pipeline is used in PulseEffects. In other to get audio from Pulseaudio the first GStreamer plugin in the chain has to be [pulsesrc](https://gstreamer.freedesktop.org/documentation/pulseaudio/pulsesrc.html?gi-language=c) and in order play audio to the soundcard the plugin [pulsesink](https://gstreamer.freedesktop.org/documentation/pulseaudio/pulsesink.html?gi-language=c) has to be used. Both of them have a `buffer size` parameter that by default has a value of `200 ms`. So the default latency is high. Of course the user can tweak the buffer size and use the smallest one possible. But depending on the output device this may not be possible. I have bluetooth headsets that force me to use buffer sizes higher than 200 ms.

There are other latency sources that are Pulseaudio only and may not be obvious to the average user. Depending on the sound card driver the default latency Pulseaudio sets to its associated sink can be high. My card is an ALC1220 and by default Pulseaudio sets its latency to 100 ms even if the optical port is being used. Fortunately we can edit a Pulseaudio config file and force it to use a value like `5 ms` that is way smaller. But I bet that most of the users have no idea this can happen. You can see this latency by running `pactl list sinks`.

That is what I was talking about. The default latency most PulseEffects users will have `by default` is way higher. Only the users with more technical knowledge will touch the knobs and configuration files that allow the latency to go to smaller values. That was what I had in mind but I did no write.

**Comment by sadko4u** [comment]
@Digitalone1 @v-fox 

I've done some review of the compressor plugin (single-band and multiband) and the final behaviour of the expermental device that I've produced in this topic.

I think I better should not build-in surge protection circuit into the compressor device for the following main reasons:
1. The device introduces huge latency when we use it with fade-outs.
2. Even if the device has 0 ms fade-out, there is an RMS value that is computed for a short period of time and used to detect proper position of the signal fall-off. So that will yield the compressor to introduce additional latency even if the fall-off surge protection is disabled.
3. Enabling surge protection for external side chain makes the things even worse and does not allow to reuse lookahead delay provided by the compressor. So for the right case latency should be result of summing compressor's lookahead delay and surge protection filter's latency.
4. We need to add at least 2 additional parameters to the UI (raise-on surge protection time, mode and threshold), at most the whole 9 parameters (all except 'Input' and 'Output' in the 'Surge Filter' group). So this will bloat the UI which is already the point of complaints from users for it's complexity.
5. Probably integrating surge protection into compressor module will require then to add the same solution to other dynamic processing plugins - Expander, Gate and Dynamic Processor.

So I don't want deliver pain to users who uses compressor mostly in downward mode with new additional chains, UI changes and latency.

If you use compressor in upward mode, then please properly set up a surge filter module. I think it is pretty ready for use.

The surge filter plugin will be available in 1.1.24 release. Currently I'll concentrate my effort on writing documentation to it.

**Comment by Digitalone1** [comment]
No worries. Setting the lookahead in sidechain resolves the issue. Don't know if surge filter is suitable for a new PE plugin, its use is very specific, most of people use only equalizer and downward compressor.

Regarding the compressor, did you change something that requires to modify existing parameters (release, attack, knee) to get the same result in the upcoming version?

**Comment by sadko4u** [comment]
@Digitalone1 no, compressor is still untouched in this branch.

**Comment by Digitalone1** [comment]
Another thing. You talked about documentation. Am I wrong or source parameters are mixed in type ones [here](https://lsp-plug.in/?page=manuals&section=sc_compressor_stereo)? 

Then I see other settings not exposed in PulseEffects like attack and release threshold, LPF and HPF. Were those added in one of the latest releases?

**Comment by sadko4u** [comment]
@Digitalone1 when talking about documenation, I was about the new plugin (each plugin should be documented).

But anyway yes, in 1.1.22 there was a change that adds Lowpass/Hipass filters to the sidechain: https://github.com/sadko4u/lsp-plugins/issues/75

And yes, it also was very painful related to the UI changes.

**Comment by sadko4u** [comment]
Merged changes into **devel** branch: b0b8f37a9ba6cace9715b34380e1d7b17aa76a17

**Comment by sadko4u** [comment]
Removed **github-issue-121** branch

**Comment by sadko4u** [comment]
Available in 1.1.24!
